### PR TITLE
Code Review for [processing-data] | DevThoughts Blog Rails Project

### PR DESCRIPTION
**Pull request by luigirazum**
# Hello Dear Code Reviewers Team 👋,
### - 🎯 General requirements
- [X] Make sure that there are _**no** linter errors_.
  - 🟢 `rubocop` ran and **no offenses** were detected.
- [X] Make sure that you used _**correct** flow_ `git-flow`.
  - 🟢 The <kbd>Milestone</kbd> branch `processing-data` was created from `dev` to follow the `git-flow`.
- [X] Make sure that you documented your work _in a **professional way**_.
  - 🟢 All the information about the project was added to the `README.md` file, previously.

### - 🎯 Project requirements
- #### Create Models
  - [X] Create model classes for all entities as shown in the **ERD diagram**.
    - 🟢 Model classes for `User`, `Post`, `Comment` and `Like` were created.
  - [X] Set up associations between models.
    _Remember that `author_id` column in `posts` table should be the foreign_key for the `users` table.__
    - 🟢 Associations among all model classes(`User`, `Post`, `Comment` and `Like`) were created, according to the **ERD diagram**.

- #### Use models to insert data
  Open rails console for your project.
  - [X] Create at least one user by running the following code:
    - 🟢 No issues creating a couple users `tom` and `lilly`.
      ```ruby
      tom = User.create(name: 'Tom', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Mexico.')
      lilly = User.create(name: 'Lilly', photo: 'https://unsplash.com/photos/F_-0BxGuVvo', bio: 'Teacher from Poland.')
      ```
  - [X] Create at least 4 posts written by one of the users you created by running the following code:
    - 🟢 `tom` created a couple `posts`.
      ```ruby
      tom_first_post = Post.create(author: tom, title: 'Hello', text: 'I\'m Tom, this is my first post')
      tom_second_post = Post.create(author: tom, title: 'Hello for 2nd time', text: 'This second post by Tom')
      tom_third_post = Post.create(author: tom, title: 'Third Hello', text: 'This is the third post from Tom')
      tom_fourth_post = Post.create(author: tom, title: 'The last Hello', text: 'This is Tom\'s fourth post')
      ```
    - 🟢 Also `lilly` created a couple `posts` as well.
      ```ruby
      lilly_first_post = Post.create(author: lilly, title: 'Lilly Hello', text: 'I\'m Lilly, a friend of Tom and this is my first post')
      lilly_second_post = Post.create(author: lilly, title: 'This is me', text: 'If you missed the first one')
      ```
  - [X] Create at least 6 posts comments for one of the posts you created by running the following code:
    - 🟢 A conversation was created between `tom` and `lilly` on `tom`'s first `post`.
      ```ruby
      Comment.create(post: tom_first_post, user: lilly, text: 'Hi Tom!')
      Comment.create(post: tom_first_post, user: tom, text: 'Hi Lilly!')
      Comment.create(post: tom_first_post, user: lilly, text: 'How you doing Tom?')
      Comment.create(post: tom_first_post, user: tom, text: 'I\'m good')
      Comment.create(post: tom_first_post, user: lilly, text: 'This is a great project!')
      Comment.create(post: tom_first_post, user: tom, text: 'Thanks, we\'re still working on it')
      ```
    - 🟢 `tom` also commented `lilly`'s first `post`.
      ```ruby
      Comment.create(post: lilly_first_post, user: tom, text: 'Hi Lilly!')
      Comment.create(post: lilly_first_post, user: lilly, text: 'Hi Tom!')
      ```
  - [X] Also you can create likes for the posts created by running the following code:
    _**This test was not included in the requirements**_
    - 🟢 `lilly` gave a 'like' to `tom`'s posts.
      ```ruby
      Like.create(post: tom_first_post, user: lilly)
      Like.create(post: tom_second_post, user: lilly)
      Like.create(post: tom_third_post, user: lilly)
      Like.create(post: tom_fourth_post, user: lilly)
      Like.create(post: tom_second_post, user: tom)
      Like.create(post: tom_fourth_post, user: tom)
      ```
    - 🟢 `tom` gave a 'like' to `lilly`'s posts.
      ```ruby
      Like.create(post: lilly_first_post, user: tom)
      Like.create(post: lilly_second_post, user: tom)
      Like.create(post: lilly_second_post, user: lilly)
      ```
- #### Create custom methods
  Your models should include some additional methods.
  - **Users**
    - [X] A method that returns the 3 most recent posts for a given user.
      - 🟢 `most_recent_posts` method was created, you can provide a _number_ to this method and it will return the _number_ of **most recent posts**, also **by default** it returns the _three_ **most recent posts**.
  - **Posts**
    - [X] A method that updates the posts counter for a user.
      - 🟢 `update_posts_counter` method was created to update `posts_counter` for the `author` of that `post`.
    - [X] A method which returns the 5 most recent comments for a given post.
      - 🟢 `most_recent_comments` method was created, you can provide a _number_ to this method and it will return the _number_ of **most recent comments**, also **by default** it returns the _five_ **most recent comments**.
  - **Comments**
    - [X] A method that updates the comments counter for a post.
      - 🟢 `update_comments_counter` method was created to update `comments_counter` for that particular `post`.
  - **Likes**
    - [X] A method that updates the likes counter for a post.
      - 🟢 `update_likes_counter` method was created to update `likes_counter` for that particular `post`.
  - [X] Go to `rails c` and check if your methods are working.
      - 🟢 `rails console` was used to check that all the created methods worked as expected.

### - 🎯 ERD Diagram
- The associations were created according to the following **ERD diagram** 👇.
  blog_app_erd_v1_1

### Luis 🙋‍♂️

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  # associations
+  has_many :posts, foreign_key: 'author_id', inverse_of: 'author'
+  has_many :likes
+  has_many :comments
+
+  def most_recent_posts(num = 3)
+    posts.limit(num).order(created_at: :desc)
+  end
```
Comment (anagudelogu) [inline app/models/user.rb:9]: - [ ] You did a great job working on this milestone! 💯 However, if you look closely, this function doesn't necessarily pick the `most_recent_posts`. It takes three `posts` and then, orders them. I suggest that you order all the posts first, and then pick the first three of that ordered list! 🤓 🔧 

 ```suggestion
  def most_recent_posts(num = 3)
    posts.order(created_at: :desc).limit(num)
  end
```

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  after_destroy :update_posts_counter
+
+  # associations
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id', inverse_of: 'posts', counter_cache: :posts_counter
+  has_many :comments
+  has_many :likes
+
+  def most_recent_comments(number = 5)
+    comments.limit(number).order(created_at: :desc)
+  end
```
Comment (anagudelogu) [inline app/models/post.rb:12]: - [ ] The same happens here, I suggest changing the order of the operations! 🤓 🔧 

```suggestion
  def most_recent_comments(number = 5)
    comments.order(created_at: :desc).limit(number)
  end
```

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS:  CHANGES REQUIRED  ♻️ 🚧 

Hi @luigirazum, 
I'm Andres.


While you have done a great job working on this milestone 💻 , there are some issues that you still need to work on to go to the next project. ♻️ 

You are almost there, keep up the good work! 🚀 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Descriptive PR title, summary and commit messages. 💬 

## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 2 above)
Comment (luigirazum) [reply app/models/post.rb:12]: Thanks for the feedback @anagudelogu.

Comment (luigirazum) [reply app/models/post.rb]: ✅ Required changes have been applied.

**Review by descholar-ceo — APPROVED** [review]
Hi @luigirazum 🤝, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
👍🏽 Implemented all requested changes ✔️ 
👍🏽 No new issues introduced while addressing ✔️ 

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**